### ISSUE TITLE
false-negative: whitespace between type annotation and array declarator or ellipsis

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17451]
split from #8205

https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s4.6.2-horizontal-whitespace
> Between a type annotation and [] or ....

A whitespace should be a blank space between a type annotation and `[]` or `...`.

```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

/** some javadoc. */
public class Test2 {
  @Target(ElementType.TYPE_USE)
  @interface Size {
    int max();
  }

  @Target(ElementType.TYPE_USE)
  @interface NonNull {}

  /** some javadoc. */
  void test1(String @NonNull... param) {}

  /** some javadoc. */
  void test2(String @NonNull[] ... param) {}

  /** some javadoc. */
  void test3(String @Size(max = 10)... names) {}
}
```

CLI:
```
$ java -jar .\checkstyle-10.23.0-all.jar -c .\google_checks.xml .\Test.java
Starting audit...
Audit done.
```

Formatter:
```
$ diff -Naru Test2.java FormattedCode.java 
--- Test2.java  2025-07-24 12:40:21.314660100 +0530
+++ FormattedCode.java  2025-07-24 12:40:47.165002100 +0530
@@ -12,12 +12,11 @@
   @interface NonNull {}
 
   /** some javadoc. */
-  void test1(String @NonNull... param) {}
+  void test1(String @NonNull ... param) {}
 
   /** some javadoc. */
-  void test2(String @NonNull[] ... param) {}
+  void test2(String @NonNull []... param) {}
 
   /** some javadoc. */
-  void test3(String @Size(max = 10)... names) {}
+  void test3(String @Size(max = 10) ... names) {}
 }
```

**Comment by Zopsss** [comment]
please also add comments for cases that are false-negative. @mohitsatr

**Comment by mohitsatr** [comment]
**Zopsss**: https://github.com/checkstyle/checkstyle/pull/17308#issuecomment-3101183605
<hr>

I think [WhitespaceAfter](https://checkstyle.org/checks/whitespace/whitespaceafter.html#WhitespaceAfter) would be a more good fit for this issue, because we have false-negatives like followings:
```java
/** some javadoc. */
public class Test {
  @Target(ElementType.TYPE_USE)
  @interface NonNull {}

  /** some javadoc. */
  void test1(String @NonNull... param) {}

  /** some javadoc. */
  void test2(String @NonNull[] ... param) {}

  /** some javadoc. */
  void test3(String @Size(max = 10)... names) {}
}
```
```
$ java -jar .\checkstyle-10.23.0-all.jar -c .\google_checks.xml .\Test.java
Starting audit...
Audit done.
```
And formatter formatted them to:
```java
$ java -jar .\google-java-format-1.28.0-all-deps.jar --replace .\Test.java
```
```
/** some javadoc. */
public class Test {
  @Target(ElementType.TYPE_USE)
  @interface NonNull {}

  /** some javadoc. */
  void test1(String @NonNull ... param) {}

  /** some javadoc. */
  void test2(String @NonNull []... param) {}

  /** some javadoc. */
  void test3(String @Size(max = 10) ... names) {}
}
```
Output of the formatted file:
```
$ java -jar .\checkstyle-10.23.0-all.jar -c .\google_checks.xml .\Test.java
Starting audit...
Audit done.
```
AST of the non-formatted file:
```
$ java -jar .\checkstyle-10.23.0-all.jar -t .\Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [2:0]
`--CLASS_DEF -> CLASS_DEF [2:0]
    |--MODIFIERS -> MODIFIERS [2:0]
    |   `--LITERAL_PUBLIC -> public [2:0]
    |--LITERAL_CLASS -> class [2:7]
    |--IDENT -> Test [2:13]
    `--OBJBLOCK -> OBJBLOCK [2:18]
        |--LCURLY -> { [2:18]
        |--ANNOTATION_DEF -> ANNOTATION_DEF [3:2]
        |   |--MODIFIERS -> MODIFIERS [3:2]
        |   |   `--ANNOTATION -> ANNOTATION [3:2]
        |   |       |--AT -> @ [3:2]
        |   |       |--IDENT -> Target [3:3]
        |   |       |--LPAREN -> ( [3:9]
        |   |       |--EXPR -> EXPR [3:21]
        |   |       |   `--DOT -> . [3:21]
        |   |       |       |--IDENT -> ElementType [3:10]
        |   |       |       `--IDENT -> TYPE_USE [3:22]
        |   |       `--RPAREN -> ) [3:30]
        |   |--AT -> @ [4:2]
        |   |--LITERAL_INTERFACE -> interface [4:3]
        |   |--IDENT -> NonNull [4:13]
        |   `--OBJBLOCK -> OBJBLOCK [4:21]
        |       |--LCURLY -> { [4:21]
        |       `--RCURLY -> } [4:22]
        |--METHOD_DEF -> METHOD_DEF [7:2]
        |   |--MODIFIERS -> MODIFIERS [7:2]
        |   |--TYPE -> TYPE [7:2]
        |   |   `--LITERAL_VOID -> void [7:2]
        |   |--IDENT -> test1 [7:7]
        |   |--LPAREN -> ( [7:12]
        |   |--PARAMETERS -> PARAMETERS [7:13]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [7:13]
        |   |       |--MODIFIERS -> MODIFIERS [7:13]
        |   |       |--TYPE -> TYPE [7:13]
        |   |       |   |--IDENT -> String [7:13]
        |   |       |   `--ANNOTATIONS -> ANNOTATIONS [7:20]
        |   |       |       `--ANNOTATION -> ANNOTATION [7:20]
        |   |       |           |--AT -> @ [7:20]
        |   |       |           `--IDENT -> NonNull [7:21]
        |   |       |--ELLIPSIS -> ... [7:28]
        |   |       `--IDENT -> param [7:32]
        |   |--RPAREN -> ) [7:37]
        |   `--SLIST -> { [7:39]
        |       `--RCURLY -> } [7:40]
        |--METHOD_DEF -> METHOD_DEF [10:2]
        |   |--MODIFIERS -> MODIFIERS [10:2]
        |   |--TYPE -> TYPE [10:2]
        |   |   `--LITERAL_VOID -> void [10:2]
        |   |--IDENT -> test2 [10:7]
        |   |--LPAREN -> ( [10:12]
        |   |--PARAMETERS -> PARAMETERS [10:13]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [10:13]
        |   |       |--MODIFIERS -> MODIFIERS [10:13]
        |   |       |--TYPE -> TYPE [10:13]
        |   |       |   |--IDENT -> String [10:13]
        |   |       |   |--ANNOTATIONS -> ANNOTATIONS [10:20]
        |   |       |   |   `--ANNOTATION -> ANNOTATION [10:20]
        |   |       |   |       |--AT -> @ [10:20]
        |   |       |   |       `--IDENT -> NonNull [10:21]
        |   |       |   `--ARRAY_DECLARATOR -> [ [10:28]
        |   |       |       `--RBRACK -> ] [10:29]
        |   |       |--ELLIPSIS -> ... [10:31]
        |   |       `--IDENT -> param [10:35]
        |   |--RPAREN -> ) [10:40]
        |   `--SLIST -> { [10:42]
        |       `--RCURLY -> } [10:43]
        |--METHOD_DEF -> METHOD_DEF [13:2]
        |   |--MODIFIERS -> MODIFIERS [13:2]
        |   |--TYPE -> TYPE [13:2]
        |   |   `--LITERAL_VOID -> void [13:2]
        |   |--IDENT -> test3 [13:7]
        |   |--LPAREN -> ( [13:12]
        |   |--PARAMETERS -> PARAMETERS [13:13]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [13:13]
        |   |       |--MODIFIERS -> MODIFIERS [13:13]
        |   |       |--TYPE -> TYPE [13:13]
        |   |       |   |--IDENT -> String [13:13]
        |   |       |   `--ANNOTATIONS -> ANNOTATIONS [13:20]
        |   |       |       `--ANNOTATION -> ANNOTATION [13:20]
        |   |       |           |--AT -> @ [13:20]
        |   |       |           |--IDENT -> Size [13:21]
        |   |       |           |--LPAREN -> ( [13:25]
        |   |       |           |--ANNOTATION_MEMBER_VALUE_PAIR -> ANNOTATION_MEMBER_VALUE_PAIR [13:26]     
        |   |       |           |   |--IDENT -> max [13:26]
        |   |       |           |   |--ASSIGN -> = [13:30]
        |   |       |           |   `--EXPR -> EXPR [13:32]
        |   |       |           |       `--NUM_INT -> 10 [13:32]
        |   |       |           `--RPAREN -> ) [13:34]
        |   |       |--ELLIPSIS -> ... [13:35]
        |   |       `--IDENT -> names [13:39]
        |   |--RPAREN -> ) [13:44]
        |   `--SLIST -> { [13:46]
        |       `--RCURLY -> } [13:47]
        `--RCURLY -> } [14:0]
```
The type annotation's AST is following ( all the methods' type annotations have this similar AST ):
```
        |   |       |--TYPE -> TYPE [13:13]
        |   |       |   |--IDENT -> String [13:13]
        |   |       |   `--ANNOTATIONS -> ANNOTATIONS [13:20]
```
So we just need to check if ANNOTATIONS' parent is TYPE? if it is then we should enforce whitespace after it. But currently the WhitespaceAfter check doesn't support ANNOTATIONS token so firstly we need to add support for it.

Adding support of a new token might be time-consuming so for now we should add // false-negative, ok until #xxxx comments.

romani mohitsatr what do you think of this approach?

**Comment by mohitsatr** [comment]
**mohitsatr**: https://github.com/checkstyle/checkstyle/pull/17308#issuecomment-3112264298
<hr>

Zopsss this approach is good. can we implement it through [NoWhitespaceAfter](https://checkstyle.org/checks/whitespace/nowhitespaceafter.html#NoWhitespaceAfter) ?

From Check's description:

> If the annotation is between the type and the array, like
`char @NotNull [] param` 
the check will skip validation for spaces.

I wonder why that nuance exists? and this is exactly the case we are trying to cover here. we can try to figure out if this check can be made aware of the spaces between type and the array declarator, and then we can use it for our purpose here with custom violation message?

This Check already supports ANNOTATIONS related token, so the implementation should take less time than doing it WhitespaceAfter's way.
https://github.com/checkstyle/checkstyle/blob/7557ec9d359021db44a75a14825e972a87090ae5/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java#L142

**Comment by mohitsatr** [comment]
@Zopsss WhitespaceAfter is not AST based check, so better not to extend it.

https://github.com/checkstyle/checkstyle/blob/35c40481a78dac174c8e36a85cd34f05ff3642b3/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java#L156-L159

We could either try https://github.com/checkstyle/checkstyle/issues/17451#issuecomment-3208988640 or make a new check.
@romani wdyt?

**Comment by romani** [comment]
it is AST based Check - https://github.com/checkstyle/checkstyle/blob/35c40481a78dac174c8e36a85cd34f05ff3642b3/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java#L104C1-L105C28
it position to right place based on AST, but then cheat with content of file to spot space, not ideal but historical implementation lasting in decades.

@mohitsatr , please consider it and sharer more detail on why it is not matching to this job.

**Comment by romani** [comment]
Fix is merged
